### PR TITLE
Added Filter

### PR DESCRIPTION
**Pull request by dlabrecq**
Added filter component and updated to latest ngx-bootstrap to take advantage of Angular 4.

Note: ng2-bootstrap was renamed ngx-bootstrap

screen shot 2017-06-13 at 1 57 18 pm

### CODE REVIEW

**Review by dtaylor113 — APPROVED** [review]

File: src/app/filters/examples/filter-example.component.ts [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+  matchesFilters(item: any, filters: Filter[]): boolean {
+    let matches = true;
+    filters.forEach((filter) => {
+      if (!this.matchesFilter(item, filter)) {
+        matches = false;
+        return false;
+      }
+    });
+    return matches;
+  }
```
Comment (jeff-phillips-18) [inline src/app/filters/examples/filter-example.component.ts]: Not sure I understand why 'matches' exists. Why set it to false just to return false?

File: src/app/filters/examples/filter-example.component.ts [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+      this.filterConfig.fields[index].queries = [
+        ...this.imageQueriesFixed,
+        this.separator,
+        ...this.imageQueries.filter((item: any) => {
+          if (item.value) {
+            return (item.value.toLowerCase().indexOf(val.toLowerCase()) > -1);
+          } else {
+            return true;
+          }
+        })
```
Comment (jeff-phillips-18) [inline src/app/filters/examples/filter-example.component.ts:305]: return !item.value || (item.value.toLowerCase().indexOf(val.toLowerCase()) > -1);

File: src/app/filters/examples/filter-example.component.ts [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+      ];
+    } else if (this.filterConfig.fields[index].id === 'weekDay') {
+      this.filterConfig.fields[index].queries = [
+        ...this.weekDayQueries.filter((item: any) => {
+          if (item.value) {
+            return (item.value.toLowerCase().indexOf(val.toLowerCase()) > -1);
+          } else {
+            return true;
+          }
```
Comment (jeff-phillips-18) [inline src/app/filters/examples/filter-example.component.ts:314]: return !item.value || (item.value.toLowerCase().indexOf(val.toLowerCase()) > -1);

File: src/app/filters/filter-fields.component.less [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    height: 20px;
+    margin-right: 5px;
+  }
+}
+
+.input-group {
+  .input-group-btn {
+    .dropdown-menu>.selected>a {
+      background-color: @color-pf-blue !important; // #0088ce
+      border-color: #0076b7 !important;
```
Comment (jeff-phillips-18) [inline src/app/filters/filter-fields.component.less]: Any better way to get around this than using !important?
Why #0076b7 and not a color from the palette?

File: src/app/filters/examples/filter-example.component.ts (hunk 1 above)
Comment (dlabrecq) [reply src/app/filters/examples/filter-example.component.ts]: That was taken verbatim from the Angular PF example. Will can clean that up.

File: src/app/filters/filter-fields.component.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  ngDoCheck(): void {
+    // Do a deep compare on config
+    if (!_.isEqual(this.config, this.prevConfig)) {
+      this.setupConfig();
+    }
+  }
+
+  setupConfig(): void {
+    this.prevConfig = _.cloneDeep(this.config);
+
```
Comment (jeff-phillips-18) [inline src/app/filters/filter-fields.component.ts]: What happens when this.config is undefined?

File: src/app/filters/filter-fields.component.less (hunk 4 above)
Comment (dlabrecq) [reply src/app/filters/filter-fields.component.less]: Again, this was taken verbatim from the Angular PF. I'm assuming there was a reason for doing that, but don't know what the issue was? Didn't want to change code without understanding it's history.

The color looks close to @color-pf-blue-400, so I'll use that instead.

File: src/app/filters/filter-fields.component.ts (hunk 5 above)
Comment (dlabrecq) [reply src/app/filters/filter-fields.component.ts]: No errors actually occur here. However, there are config properties that are undefined during rendering. 

That said, I've added the '?' operator in the HTML templates to allow the component to be rendered without the config. Doesn't look pretty, but no errors are output.

**Review by jeff-phillips-18 — APPROVED** [review]

**Pull request merged**